File: WordTarget/WordTargetCoreTests/LinkTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace WordTargetCore
{
    [TestClass]
    public class LinkTest
    {

        static Word cat = new Word("cat");
        static Word car = new Word("car");
        static Word cart = new Word("cart");
        static Word bat = new Word("bat");
        static Word bit = new Word("bit");
        static Word brat = new Word("brat");
        static Word tab = new Word("tab");

        [TestMethod]
        public void LinkCreation()
        {
            
            Link link = new Link(cat, bat, LinkType.OneLetterChange);
            Assert.AreEqual(cat, link.WordA);
            Assert.AreEqual(bat, link.WordB);
            Assert.AreEqual(LinkType.OneLetterChange, link.Type);
        }

        [TestMethod]
        public void VerifyAnagram()
        {
            Link link = new Link(bat, tab, LinkType.Anagram);
            Assert.ThrowsException<System.Exception>(() => new Link(cat, bat, LinkType.Anagram));
        }


        [TestMethod]
        public void VerifyOneLetterChange()
        {
            Link link = new Link(cat, bat, LinkType.OneLetterChange);
            Assert.ThrowsException<System.Exception>(() => new Link(cat, brat, LinkType.OneLetterChange));
            Assert.ThrowsException<System.Exception>(() => new Link(cat, bit, LinkType.OneLetterChange));
            Assert.ThrowsException<System.Exception>(() => new Link(cat, cat, LinkType.OneLetterChange));
        }

        [TestMethod]
        public void VerifyOneLetterAdd()
        {
            Link link = new Link(cat, cart, LinkType.OneLetterAddOrRemove);
            Assert.ThrowsException<System.Exception>(() => new Link(cat, brat, LinkType.OneLetterAddOrRemove));
            Assert.ThrowsException<System.Exception>(() => new Link(cat, bit, LinkType.OneLetterAddOrRemove));
            Assert.ThrowsException<System.Exception>(() => new Link(cat, cat, LinkType.OneLetterAddOrRemove));
        }
        
        [TestMethod]
        public void VerifyOneLetterRemove()
        {
            Link link = new Link(cart, cat, LinkType.OneLetterAddOrRemove);
            Assert.ThrowsException<System.Exception>(() => new Link(cat, brat, LinkType.OneLetterAddOrRemove));
            Assert.ThrowsException<System.Exception>(() => new Link(cat, bit, LinkType.OneLetterAddOrRemove));
            Assert.ThrowsException<System.Exception>(() => new Link(cat, cat, LinkType.OneLetterAddOrRemove));
        }

        [TestMethod]
        public void VerifyOneLetterAddOrRemove()
        {
            Link link1 = new Link(cat, cart, LinkType.OneLetterAddOrRemove);
            Link link2 = new Link(cart, cat, LinkType.OneLetterAddOrRemove);
            Assert.ThrowsException<System.Exception>(() => new Link(cat, brat, LinkType.OneLetterAddOrRemove));
            Assert.ThrowsException<System.Exception>(() => new Link(cat, car, LinkType.OneLetterAddOrRemove));
            Assert.ThrowsException<System.Exception>(() => new Link(cat, cat, LinkType.OneLetterAddOrRemove));

        }

        [TestMethod]
        public void LinkStingConversion()
        {
            Link link = new Link(cat, bat, LinkType.OneLetterChange);
            Assert.AreEqual("CAT-BAT", link.ToString());
        }

        [TestMethod]
        public void ValidLinkEquivalency()
        {
            Link link1 = new Link(cat, bat, LinkType.OneLetterChange);
            Link link2 = new Link(cat, bat, LinkType.OneLetterChange);
            Assert.AreEqual(link1, link2);
            Assert.AreEqual(link1.GetHashCode(), link2.GetHashCode());
        }

        [TestMethod]
        public void ValidLinkInversionForOneLetterChange()
        {
            Link link1 = new Link(cat, bat, LinkType.OneLetterChange);
            Link link2 = link1.Reverse();
            Assert.AreEqual(cat, link1.WordA);
            Assert.AreEqual(bat, link1.WordB);
            Assert.AreEqual(bat, link2.WordA);
            Assert.AreEqual(cat, link2.WordB);
        }
    }
}

File: word-target/word-target-core/src/test/java/it/carcassi/wordtarget/core/LinkTest.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package it.carcassi.wordtarget.core;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import static it.carcassi.wordtarget.core.CommonTestObjects.*;

/**
 *
 * @author Matteo
 */
public class LinkTest {

    @Test
    public void testLinkCreation() {

        Link link = new Link(cat, bat, LinkType.OneLetterChange);
        assertEquals(cat, link.getWordA());
        assertEquals(bat, link.getWordB());
        assertEquals(LinkType.OneLetterChange, link.getType());
    }

    @Test
    public void testVerifyAnagram() {
        Link link = new Link(bat, tab, LinkType.Anagram);
        assertThrows(IllegalArgumentException.class, () -> {
            new Link(cat, bat, LinkType.Anagram);
        });
    }

    @Test
    public void testVerifyOneLetterChange() {
        Link link = new Link(cat, bat, LinkType.OneLetterChange);
        assertThrows(IllegalArgumentException.class, () -> {
            new Link(cat, brat, LinkType.OneLetterChange);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            new Link(cat, bit, LinkType.OneLetterChange);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            new Link(cat, cat, LinkType.OneLetterChange);
        });
    }

    @Test
    public void testVerifyOneLetterAddOrRemove() {
        Link link1 = new Link(cat, cart, LinkType.OneLetterAddOrRemove);
        Link link2 = new Link(cart, cat, LinkType.OneLetterAddOrRemove);
        assertThrows(IllegalArgumentException.class, () -> {
            new Link(cat, brat, LinkType.OneLetterAddOrRemove);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            new Link(cat, car, LinkType.OneLetterAddOrRemove);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            new Link(cat, cat, LinkType.OneLetterAddOrRemove);
        });
    }

    @Test
    public void testLinkStingConversion() {
        Link link = new Link(cat, bat, LinkType.OneLetterChange);
        assertEquals("CAT-BAT", link.toString());
    }

    @Test
    public void testValidLinkEquivalency() {
        Link link1 = new Link(cat, bat, LinkType.OneLetterChange);
        Link link2 = new Link(cat, bat, LinkType.OneLetterChange);
        assertEquals(link1, link2);
        assertEquals(link1.hashCode(), link2.hashCode());
    }

    @Test
    public void testValidLinkInversionForOneLetterChange() {
        Link link1 = new Link(cat, bat, LinkType.OneLetterChange);
        Link link2 = link1.reverse();
        assertEquals(cat, link1.getWordA());
        assertEquals(bat, link1.getWordB());
        assertEquals(bat, link2.getWordA());
        assertEquals(cat, link2.getWordB());
    }
    
//    @Test
//    public void testLinkOf1() {
//        Link link = Link.of("CAT BAT OneLetterChange");
//        assertEquals(cat, link.getWordA());
//        assertEquals(bat, link.getWordA());
//        assertEquals(LinkType.OneLetterChange, link.getType());
//    }
//    @Test
//    public void testLinkOf2() {
//        Link link = Link.of("CAT 'BAT' OneLetterChange");
//        assertEquals(cat, link.getWordA());
//        assertEquals(bat, link.getWordA());
//        assertEquals(LinkType.OneLetterChange, link.getType());
//    }
//    @Test
//    public void testLinkOf3() {
//        Link link = Link.of("SIGHT 'VITAMIN A' WordAssociation");
//        assertEquals(Word.of("SIGHT"), link.getWordA());
//        assertEquals(Word.of("VITAMIN A"), link.getWordA());
//        assertEquals(LinkType.WordAssociation, link.getType());
//    }
}

